### ISSUE TITLE
Softmax derivative is wrong

### ISSUE THREAD

**Issue by giladfrid009** [void-intelligence/Vortex#11]
The softmax derivative implemented is wrong, since each input of softmax layer affects other inputs of the same layer. Therefore, when calculating the derivative of certain softmax input, other inputs have to be taken into account also.
A good explanation and clarification on the topic could be found here: (check Viviek's answer) https://stats.stackexchange.com/questions/235528/backpropagation-with-softmax-cross-entropy?newreg=2a0e1acc95e345d6b1d0197d64fa1b25

Xi - input value of index i
S(Xi) - softmax of Xi

Softmax derivative when `index input(i) == index output(j):` `S(Xi) * (1-S(Xj))`

Softmax derivative when `index input (i) != index output (j):` `-S(Xi) * S(Xj)`

You can check my custom softmax layer implementation for more details.
